File: Homework_01/container.cpp
// Copyright 2021, German Mikhaylov.
//
// This file is part of Homework_01.
//
//    Homework_01 is free software: you can redistribute it and/or modify
//    it under the terms of the GNU General Public License as published by
//    the Free Software Foundation, either version 3 of the License, or
//    (at your option) any later version.
//
//    Homework_01 is distributed in the hope that it will be useful,
//    but WITHOUT ANY WARRANTY; without even the implied warranty of
//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//    GNU General Public License for more details.
//
//    You should have received a copy of the GNU General Public License
//    along with Homework_01.  If not, see <https://www.gnu.org/licenses/>.
//
//-------------------------------------------------------------------------

/**
 * container.cpp - содержит функции обработки контейнера
 */

#include "container.h"

// Инициализация контейнера
void Init(container &c) {
    c.len = 0;
}

// Очистка контейнера от элементов (освобождение памяти)
void Clear(container &c) {
    c.len = 0;
}

// Ввод содержимого контейнера из указанного потока
void In(container &c, std::ifstream &ifst) {
    while (!ifst.eof()) {
        if (In(c.container[c.len], ifst)) {
            c.len++;
        }
    }
}

// Случайный ввод содержимого контейнера
void InRandom(container &c, int size) {
    while (c.len < size) {
        if (InRandom(c.container[c.len])) {
            c.len++;
        }
    }
}

// Вывод содержимого контейнера в указанный поток
void Out(container &c, std::ofstream &ofst) {
    ofst << "Container contains " << c.len << " elements.\n" << std::endl;
    for (int i = 0; i < c.len; i++) {
        ofst << i << ": ";
        Out(c.container[i], ofst);
    }
}

// Функция удаления матриц, если ср. арифметическое элементов меньше среднего в контейнере
void RemoveByAverage(container &c) {
    double sum = 0.0;
    for (int i = 0; i < c.len; ++i) {
        sum += Average(c.container[i]);
    }
    double containerAverage = sum / c.len;

    matrix containerAfterDelete[c.len];
    int pointer = 0;
    for (int i = 0; i < c.len; ++i) {
        if (Average(c.container[i]) >= containerAverage) {
            containerAfterDelete[pointer] = c.container[i];
            ++pointer;
        }
    }
    c.len = pointer;

    for (int i = 0; i < c.len; ++i) {
        c.container[i] = containerAfterDelete[i];
    }
}

File: Homework_03/container.py
""" Copyright 2021, German Mikhaylov.

 This file is part of Homework_03.

    Homework_03 is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    Homework_03 is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Homework_03.  If not, see <https://www.gnu.org/licenses/>.

-------------------------------------------------------------------------"""

from DiagonalMatrix import DiagonalMatrix
from RegularMatrix import RegularMatrix
from TriangleMatrix import TriangleMatrix
import random


class Container:
    def __init__(self):
        self.store = []

    def print_container(self):
        print("Container is store", len(self.store), "matrix:")
        for matrix in self.store:
            matrix.print_to_console()
        pass

    def write_container(self, output_stream):
        output_stream.write("Container is store {} matrix:\n".format(len(self.store)))
        for matrix in self.store:
            matrix.write_to_file(output_stream)
            output_stream.write("\n")
        pass

    def remove_by_average(self):
        summa = 0.0
        length = len(self.store)
        for i in range(length):
            summa += self.store[i].get_average()

        container_average = summa / length

        container_after_delete = []
        for i in range(len(self.store)):
            if self.store[i].get_average() >= container_average:
                container_after_delete.append(self.store[i])

        length = len(container_after_delete)
        self.store.clear()

        for i in range(length):
            self.store.append(container_after_delete[i])
        pass

    def read_str_array(self, str_array):
        array_len = len(str_array)
        # Индекс, задающий текущую строку в массиве
        i = 0
        fig_num = 0
        while i < array_len:
            key = int(str_array[i])
            if key == 1:
                i += 1
                matrix = RegularMatrix()
                # чтение обычной матрицы с возвратом позиции за ним
                i = matrix.read_from_file(str_array, i)
            elif key == 2:
                i += 1
                matrix = DiagonalMatrix()
                # чтение диагональной матрицы с возвратом позиции за ним
                i = matrix.read_from_file(str_array, i)
            elif key == 3:
                i += 1
                matrix = TriangleMatrix()
                # чтение треугольной матрицы с возвратом позиции за ним
                i = matrix.read_from_file(str_array, i)
            else:
                # Неверный признак
                # Возврат количества прочитанных фигур
                return fig_num

            fig_num += 1
            self.store.append(matrix)
        return fig_num

    def create_random(self, number):
        for i in range(number):
            self.store.append(random_matrix())
        pass


def random_matrix():
    identifier = random.randint(1, 3)
    matrix = None
    if identifier == 1:
        matrix = RegularMatrix()
    elif identifier == 2:
        matrix = DiagonalMatrix()
    elif identifier == 3:
        matrix = TriangleMatrix()

    matrix.random_input()
    return matrix
